controllers/util/conditions.go

```diff
@@ -87,28 +88,108 @@ func ConditionAppend(
 	)
 }
 
-func ConditionSetFirstFalseOrLastTrue(
+// MergeConditions merges VRG conditions of the same type to generate a single condition for the Type
+func MergeConditions(
 	conditionSet func(*[]metav1.Condition, metav1.Condition),
 	conditions *[]metav1.Condition,
+	ignoreReasons []string,
 	subConditions ...*metav1.Condition,
 ) {
-	trueSubConditions := make([]*metav1.Condition, 0, len(subConditions))
+	trueSubConditions := []*metav1.Condition{}
+	falseSubConditions := []*metav1.Condition{}
+	unknownSubConditions := []*metav1.Condition{}
 
 	for _, subCondition := range subConditions {
 		if subCondition == nil {
 			continue
 		}
 
-		if subCondition.Status == metav1.ConditionFalse {
-			conditionSet(conditions, *subCondition)
-
-			return
+		switch subCondition.Status {
+		case metav1.ConditionFalse:
+			falseSubConditions = oldestConditions(subCondition, falseSubConditions)
+		case metav1.ConditionUnknown:
+			unknownSubConditions = oldestConditions(subCondition, unknownSubConditions)
+		case metav1.ConditionTrue:
+			trueSubConditions = oldestConditions(subCondition, trueSubConditions)
 		}
+	}
+
+	switch {
+	case len(falseSubConditions) != 0:
+		conditionSet(conditions, mergedCondition(falseSubConditions, ignoreReasons))
+	case len(unknownSubConditions) != 0:
+		conditionSet(conditions, mergedCondition(unknownSubConditions, ignoreReasons))
+	case len(trueSubConditions) != 0:
+		conditionSet(conditions, mergedCondition(trueSubConditions, ignoreReasons))
+	}
+}
+
+// oldestConditions returns a list of conditions that are the same generation and the oldest among newCondition and
+// conditions[]. Initial call should pass an empty conditions[] and iterate over conditions to merge, as the function
+// expects conditions in passed in conditions[] to be the of the same generation.
+func oldestConditions(newCondition *metav1.Condition, conditions []*metav1.Condition) []*metav1.Condition {
+	retConditions := []*metav1.Condition{}
+
+	if len(conditions) == 0 {
+		retConditions = append(retConditions, newCondition)
```

Comment:
(nit) maybe just call to eliminate the 2 lines and reduce typing mistakes between `conditions` and `retConditions`:
```
return append(retConditions, newCondition)
```